src/printer.js

```diff
@@ -2522,13 +2522,20 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
 
         if (/\S/.test(value)) {
           // treat each line of text as its own entity
-          value.split(/(\n\s*)/).forEach(line => {
+          const lines = value.split(/(\n\s*)/);
+
+          // if there's a single line of text, we don't want to force a
+          // hardline but make it fit in a single line if possible
+          const isSingleText = n.children.length === 1 &&
+            lines.filter(l => l.trim()).length === 1;
```

Comment:
`l => /\S/.test(l)` might be faster (and possibly more clear) but probably doesn't matter.